ta/pkcs11/src/object.c

```diff
@@ -0,0 +1,408 @@
+ * @handle - generated handle for the created object
+ */
+enum pkcs11_rc create_object(void *sess, struct obj_attrs *head,
+			     uint32_t *out_handle)
+{
+	enum pkcs11_rc rc = 0;
+	struct pkcs11_object *obj = NULL;
+	struct pkcs11_session *session = (struct pkcs11_session *)sess;
+	uint32_t obj_handle = 0;
+
+#ifdef DEBUG
+	trace_attributes("[create]", head);
+#endif
+
+	/*
+	 * We do not check the key attributes. At this point, key attributes
+	 * are expected consistent and reliable.
+	 */
+
+	obj = create_obj_instance(head);
+	if (!obj)
+		return PKCS11_CKR_DEVICE_MEMORY;
+
+	/* Create a handle for the object in the session database */
+	obj_handle = handle_get(&session->object_handle_db, obj);
+	if (!obj_handle) {
+		rc = PKCS11_CKR_DEVICE_MEMORY;
+		goto err;
+	}
+
+	if (get_bool(obj->attributes, PKCS11_CKA_TOKEN)) {
+		TEE_Result res = TEE_SUCCESS;
+
+		/*
+		 * Get an ID for the persistent object
+		 * Create the file
+		 * Register the object in the persistent database
+		 * (move the full sequence to persisent_db.c?)
+		 */
+		size_t size = sizeof(struct obj_attrs) +
+			      obj->attributes->attrs_size;
+		uint32_t tee_obj_flags = TEE_DATA_FLAG_ACCESS_READ |
+					 TEE_DATA_FLAG_ACCESS_WRITE |
+					 TEE_DATA_FLAG_ACCESS_WRITE_META;
+
+		rc = create_object_uuid(get_session_token(session), obj);
+		if (rc)
+			goto err;
+
+		res = TEE_CreatePersistentObject(TEE_STORAGE_PRIVATE,
+						 obj->uuid, sizeof(TEE_UUID),
+						 tee_obj_flags,
+						 TEE_HANDLE_NULL,
+						 obj->attributes, size,
+						 &obj->attribs_hdl);
+		if (res) {
+			rc = tee2pkcs_error(res);
+			goto err;
+		}
+
+		rc = register_persistent_object(get_session_token(session),
+						obj->uuid);
+		if (rc)
+			goto err;
+
+		LIST_INSERT_HEAD(&session->token->object_list, obj, link);
+	} else {
+		rc = PKCS11_CKR_OK;
+		LIST_INSERT_HEAD(get_session_objects(session), obj, link);
+	}
+
+	*out_handle = obj_handle;
+
+	return PKCS11_CKR_OK;
+err:
+	/* make sure that supplied "head" isn't freed */
+	obj->attributes = NULL;
+	handle_put(&session->object_handle_db, obj_handle);
+	if (get_bool(head, PKCS11_CKA_TOKEN))
+		cleanup_persistent_object(obj, session->token);
+	else
+		cleanup_volatile_obj_ref(obj);
+
+	return rc;
+}
+
+enum pkcs11_rc entry_import_object(struct pkcs11_client *client,
+				   uint32_t ptypes, TEE_Param *params)
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_MEMREF_OUTPUT,
+						TEE_PARAM_TYPE_NONE);
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+	TEE_Param *ctrl = params;
+	TEE_Param *out = params + 2;
+	struct serialargs ctrlargs = { };
+	struct pkcs11_session *session = NULL;
+	struct obj_attrs *head = NULL;
+	struct pkcs11_object_head *template = NULL;
+	uint32_t session_handle = 0;
+	size_t template_size = 0;
+	uint32_t obj_handle = 0;
+
+	/*
+	 * Collect the arguments of the request
+	 */
+
+	if (!client || ptypes != exp_pt ||
+	    out->memref.size != sizeof(obj_handle))
```

Comment:
While this is correct, it might be a little to defensive. `< ` instead of `!=` would do here, as the size is set in line #351 anyway.

Reply:
There's nothing to be gained from being flexible here, if the caller expects a handle of a different size it's beyond rescue anyway.

Reply:
Ok.